packages/react-query/src/errorBoundaryUtils.ts

```diff
@@ -0,0 +1,67 @@
+import {
+  DefaultedQueryObserverOptions,
+  Query,
+  QueryKey,
+  QueryObserverResult,
+} from '@tanstack/query-core'
+import { QueryErrorResetBoundaryValue } from './QueryErrorResetBoundary'
+import * as React from 'react'
+import { shouldThrowError } from './utils'
+
+export const ensurePreventErrorBoundaryRetry = <
+  TQueryFnData,
+  TError,
+  TData,
+  TQueryData,
+  TQueryKey extends QueryKey,
+>(
+  options: DefaultedQueryObserverOptions<
+    TQueryFnData,
+    TError,
+    TData,
+    TQueryData,
+    TQueryKey
+  >,
+  errorResetBoundary: QueryErrorResetBoundaryValue,
+) => {
+  if (options.suspense || options.useErrorBoundary) {
+    // Prevent retrying failed query if the error boundary has not been reset yet
+    if (!errorResetBoundary.isReset()) {
+      options.retryOnMount = false
+    }
+  }
+}
+
+export const useClearResetErrorBoundary = (
+  errorResetBoundary: QueryErrorResetBoundaryValue,
+) => {
+  React.useEffect(() => {
+    errorResetBoundary.clearReset()
+  }, [errorResetBoundary])
+}
+
+export const getHasError = <
+  TData,
+  TError,
+  TQueryFnData,
+  TQueryData,
+  TQueryKey extends QueryKey,
+  TUseErrorBoundaryFn extends (...args: any[]) => boolean,
+>({
+  result,
+  errorResetBoundary,
+  useErrorBoundary,
+  query,
+}: {
+  result: QueryObserverResult<TData, TError>
+  errorResetBoundary: QueryErrorResetBoundaryValue
+  useErrorBoundary: boolean | TUseErrorBoundaryFn | undefined
+  query: Query<TQueryFnData, TError, TQueryData, TQueryKey>
+}) => {
+  return (
+    result.isError &&
+    !errorResetBoundary.isReset() &&
+    !result.isFetching &&
+    shouldThrowError(useErrorBoundary, [result.error, query] as Parameters<TUseErrorBoundaryFn>)
```

Comment:
Wonder if this `as Parameters<TUseErrorBoundaryFn>` could be avoided somehow

Reply:
it works if we hardcode the type of `useErrorBoundary` without the extra generic:

```
useErrorBoundary: boolean | undefined | ((
    e: TError,
    q: Query<TQueryFnData, TError, TQueryData, TQueryKey>,
  ) => boolean)
```

maybe it would just be good to extract this type:

https://github.com/TanStack/query/blob/f828528c935089a3adbf7b0474c6c91603ff47e4/packages/query-core/src/types.ts#L208-L213

and then re-use it here?

Reply:
Thanks, I'll try it out.

Reply:
[3518f3c](https://github.com/TanStack/query/pull/4177/commits/3518f3ca403d52c8be23073c9d66af8a250e34b8)